math/dec.go

```diff
@@ -368,12 +368,30 @@ func (x Dec) Reduce() (Dec, int) {
 
 // Marshal serializes the decimal value into a byte slice in a text format.
 // This method ensures the decimal is represented in a portable and human-readable form.
-func (d Dec) Marshal() ([]byte, error) {
-	return d.dec.MarshalText()
+// The output may be in scientific notation if the number's magnitude is very large or very small.
+//
+// Returns:
+// - A byte slice of the decimal in text format, which may include scientific notation depending on the value.
+func (x Dec) Marshal() ([]byte, error) {
+	return x.dec.MarshalText()
 }
 
 // Unmarshal parses a byte slice containing a text-formatted decimal and stores the result in the receiver.
 // It returns an error if the byte slice does not represent a valid decimal.
-func (d *Dec) Unmarshal(data []byte) error {
-	return d.dec.UnmarshalText(data)
+func (x *Dec) Unmarshal(data []byte) error {
+	var d apd.Decimal
+	_, _, err := d.SetString(string(data))
+	if err != nil {
+		return ErrInvalidDec.Wrap(err.Error())
+	}
+
+	switch d.Form {
+	case apd.NaN, apd.NaNSignaling:
+		return ErrInvalidDec.Wrap("not a number")
+	case apd.Infinite:
+		return ErrInvalidDec.Wrap("infinite decimal value not allowed")
+	default:
```

Comment:
The success case should probably be `apd.Finite`, with `default:` actually being an error for unknown form